pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	udpa "github.com/cncf/udpa/go/udpa/type/v1"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	hcm_filter "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"google.golang.org/protobuf/proto"
+
+	extensions "istio.io/api/extensions/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking"
+	"istio.io/istio/pilot/pkg/networking/util"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	securitymodel "istio.io/istio/pilot/pkg/security/model"
+)
+
+const (
+	wasmFilterType = "envoy.extensions.filters.http.wasm.v3.Wasm"
+)
+
+var defaultConfigSource = &envoy_config_core_v3.ConfigSource{
+	ConfigSourceSpecifier: &envoy_config_core_v3.ConfigSource_Ads{
+		Ads: &envoy_config_core_v3.AggregatedConfigSource{},
+	},
+	ResourceApiVersion: envoy_config_core_v3.ApiVersion_V3,
+}
+
+// AddWasmPluginsToMutableObjects adds WasmPlugins to HTTP filterChains
+// Note that the slices in the map must already be ordered by plugin
+// priority! This will be the case for maps returned by PushContext.WasmPlugin()
+func AddWasmPluginsToMutableObjects(
+	mutable *networking.MutableObjects,
+	extensionsMap map[extensions.PluginPhase][]*model.WasmPluginWrapper,
+) {
+	if mutable == nil {
+		return
+	}
+
+	for fcIndex, fc := range mutable.FilterChains {
+		// we currently only support HTTP
+		if fc.ListenerProtocol != networking.ListenerProtocolHTTP {
+			continue
+		}
+		mutable.FilterChains[fcIndex].HTTP = injectExtensions(fc.HTTP, extensionsMap)
+	}
+}
+
+func injectExtensions(filterChain []*hcm_filter.HttpFilter, exts map[extensions.PluginPhase][]*model.WasmPluginWrapper) []*hcm_filter.HttpFilter {
+	// copy map as we'll manipulate it in the loop
+	extMap := make(map[extensions.PluginPhase][]*model.WasmPluginWrapper)
+	for phase, list := range exts {
```

Comment:
Should we define priorities for built in filters, so that Wasm filters can be injected after the built in ones, not always before them?

Reply:
injecting after a builtin filter can also be done my using the next phase, with a high priority. do you think that's enough?

Reply:
This is just a random idea coming from the ext authz filter ordering comment, so that we can get complete control over ordering of built-in & Wasm filters if we have predefined priority of built in filters. We discussed this a bit at E&T WG meeting and decided this can save to later iteration.